Source/WebCore/loader/ContentFilter.cpp

```diff
@@ -123,6 +123,76 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
     return !request.isNull();
 }
 
+ContentFilter::ContentFilterCallbackAggregator::~ContentFilterCallbackAggregator()
+{
+    ASSERT(RunLoop::isMain());
+    ASSERT(m_callback);
+
+    if (m_isBlocked) {
+        if (CheckedPtr contentFilter = m_contentFilter)
+            contentFilter->didDecide(State::Blocked);
+        m_callback(ResourceRequest());
+        return;
+    }
+
+    if (CheckedPtr contentFilter = m_contentFilter; m_numberOfFiltersAllowed == contentFilter->m_contentFilters.size())
+        contentFilter->didDecide(State::Allowed);
+
+    m_callback(WTFMove(m_request));
+}
+
+void ContentFilter::ContentFilterCallbackAggregator::didReceivePlatformContentFilterDecision(PlatformContentFilter& platformContentFilter, ResourceRequest&& request, PlatformContentFilter::State state)
+{
+    ASSERT(RunLoop::isMain());
+
+    if (state == PlatformContentFilter::State::Allowed)
+        ++m_numberOfFiltersAllowed;
+
+    bool isBlocked = state == PlatformContentFilter::State::Blocked;
+    if (isBlocked && !m_isBlocked) {
+        m_isBlocked = true;
+        CheckedPtr contentFilter = m_contentFilter;
+        contentFilter->m_blockingContentFilter = platformContentFilter;
+        return;
+    }
+
+    m_request = WTFMove(request);
+}
+
+ContentFilter::ContentFilterCallbackAggregator::ContentFilterCallbackAggregator(ContentFilter& contentFilter, CompletionHandler<void(ResourceRequest&&)>&& callback)
+    : m_contentFilter(contentFilter)
+    , m_callback(WTFMove(callback))
+{
+    ASSERT(RunLoop::isMain());
+}
+
+void ContentFilter::continueAfterWillSendRequest(ResourceRequest&& request, const ResourceResponse& redirectResponse, CompletionHandler<void(ResourceRequest&&)>&& completionHandler)
+{
+    ASSERT(RunLoop::isMain());
+
+    // Return early if no platform content filters are enabled.
+    if (!anyContentFilterIsEnabled()) {
```

Comment:
Is this check necessary? If no content filter is enabled, shouldn't `PlatformContentFilter` directly invoke completionHandler with allowed?

Reply:
That's a good point! I'll address this issue and the other comments shortly.

Thanks for reviewing!